### DISCUSSION TITLE
[Bug] `Variables.Add` ignores `validator` and `converter` if `help` and `default` weren't passed

### DISCUSSION THREAD

**Discussion by ghost** [SCons/scons#4652] [Q&A] [answered]
Look at following file:

```python
import SCons
from SCons.Variables import *
env = Environment()

options = Variables()

def validator(key, val, env):
	raise SCons.Errors.UserError(f"validator: {key} {val}") from None

def converter(val):
	return False

options.Add("my_option", validator=validator, converter=converter)

options.Update(env)

print(options.GenerateHelpText(env))

print(options)
```

There is only option name passed aside converter and validator. Run `scons my_option="no". The code prints following:

```
scons: Reading SConscript files ...

my_option: 
    default: None
    actual: None

Variables(
  options=[
    ('my_option', ['my_option'], '', None, validator=<function validator at 0x7f09b02a5a80>, converter=<function converter at 0x7f09b02ef920>),
  ],
  args={},
  files=[],
  unknown={},
)
```

Here we see that it ignored validator and converter but if you will change the option from

```diff
- options.Add("my_option", validator=validator, converter=converter)
+ options.Add("my_option", "help", "default", validator=validator, converter=converter)
```

It will print as it should `scons: *** validator: my_option False`... The same is with converter...

Is it known? Should I convert it to issue?

**Comment by mwichmann** [comment]
Due to lack of a default, meaning `default` defaults to `None`, which means it isn't added to the initial dict which apparently messes up later processing. Looks like a bug, let me give it a bit more thought.

**Reply by ghost** [reply]
Ok. I've just tested and explicitly passed `None` as `default` and as a result it is still not expected behavior. If users case defaults to `None` explicitly (which is fairly possible) means that `validator` should work with `None` and if it can't it is a bug

**Comment by mwichmann** [comment, marked as answer]
This didn't turn out to be what I expected.  Remove the star import from `Variables` - the publicly exposed (from an SConscript viewpoint) `Variables` is a function in the main part of the program (`SCons.Script`), which instantiates the `Variables` class with appropriate defaults.  The star import overwrites that function with the symbol (class name) from `SCons.Variables`, so you get the wrong one, and in fact, the Variables package doesn't end up seeing your command-line argument.

**Reply by ghost** [reply]
Thanks!
